[GStreamer][EME] Break key waiting on PAUSED->READY state change

calvaris opened the pull request:
#### a65d75e0fad55dedd44857659f21bd7ee81b4e1e
<pre>
[GStreamer][EME] Break key waiting on PAUSED-&gt;READY state change
<a href="https://bugs.webkit.org/show_bug.cgi?id=258070">https://bugs.webkit.org/show_bug.cgi?id=258070</a>

Reviewed by Philippe Normand.

Changing a state can&apos;t be completed when decryptor is waiting for a key.
This can block main thread for max 7sec.

1) Signal key waiting condition to return when decryptor is closing.
2) Return waiting for CDMProxy for the same case.

Original patch by: Andrzej Surdej &lt;Andrzej_Surdej@comcast.com&gt;.

* LayoutTests/media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key-expected.txt: Added.
* LayoutTests/media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key.html: Added.
* LayoutTests/media/encrypted-media/clearKey/encrypted-media-clearKey-handler.js:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
(CDMProxyDecryptionClientImplementation::isAborting):
(transformInPlace):
(webKitMediaCommonEncryptionDecryptIsAborting):
(changeState):
(webKitMediaCommonEncryptionDecryptIsFlushing): Deleted.
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/265231@main">https://commits.webkit.org/265231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/384ad08b2877796ca124ed7ba40b751874fa2e17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9839 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10212 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12771 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10358 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11101 "3 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8598 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12229 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8406 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9237 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16524 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9505 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12630 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9847 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7993 "1 flakes 4 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9006 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9132 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2474 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9674 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
```diff
@@ -482,10 +487,23 @@ static GstStateChangeReturn changeState(GstElement* element, GstStateChange tran
+        priv->isStopped = false;
+        break;
+    }
+    case GST_STATE_CHANGE_PAUSED_TO_READY: {
```
This should go to the `// Add post-transition code here.` section a couple lines below.

philn commented on LayoutTests/media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key.html:
```diff
@@ -0,0 +1,45 @@
+                 const startTime = Date.now();
+                 video.src = "../../content/counting.mp4"
+                 video.play(1);
+                 const difference = Date.now() - startTime;
```
let's hope this won't be a source of flakiness on our constantly overloaded bots...

calvaris replied:
It usually takes 2 ms if test works but I can even increase it a bit more to reduce the possibilities of flakyness.

philn replied:
Found 1 new test failure: media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key.html

calvaris replied:
Yes, it's because moving handling the state change after calling the superclass. That does not break the wait. Added a comment about it. Also added skip for mac.

calvaris committed:
[GStreamer][EME] Break key waiting on PAUSED->READY state change
https://bugs.webkit.org/show_bug.cgi?id=258070

Reviewed by Philippe Normand.

Changing a state can't be completed when decryptor is waiting for a key.
This can block main thread for max 7sec.

1) Signal key waiting condition to return when decryptor is closing.
2) Return waiting for CDMProxy for the same case.

Original patch by: Andrzej Surdej <Andrzej_Surdej@comcast.com>.

* LayoutTests/media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key-expected.txt: Added.
* LayoutTests/media/encrypted-media/clearKey/clearKey-cenc-stop-playback-before-key.html: Added.
* LayoutTests/media/encrypted-media/clearKey/encrypted-media-clearKey-handler.js:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
(CDMProxyDecryptionClientImplementation::isAborting):
(transformInPlace):
(webKitMediaCommonEncryptionDecryptIsAborting):
(changeState):
(webKitMediaCommonEncryptionDecryptIsFlushing): Deleted.
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.h:

Canonical link: https://commits.webkit.org/265231@main
